### DISCUSSION TITLE
Memory usage 2D grid lookup grows quadratically

### DISCUSSION THREAD

**Discussion by jwmuller-tue** [jax-ml/jax#40983] [Q&A]
I want to interpolate an input_data with shape [Nchannel, Ndatapoints], where the resulting output is a 2D grid which is defined using two 1D arrays "x" and "z". The input is sum-reduced over the Nchannel axis, and there is a non-linear interaction between x, y and an offset per Nchannel axis to determine which entry in the Ndatapoints axis of input_data to obtain. The nearest neighbor interpolation memory only scales linearly when both "x" and "y" are increased linearly. For linear interpolation the increase is quadratic. It seems that for nearest neighbor interpolation, 2 intermediate tensors are created for the x and y data separately. For linear interpolation, the intermediate tensor is first expanded to a full x-y 2D grid before the interpolation step, which explains the quadratic growth of the memory.

Is there any method I can apply to reduce the memory usage for linear interpolation? Eventually I want to use much larger grid sizes as well as gradients, where the problem is even worse. Scanning over the channels is really slow and I also tried to change the order of the vmaps with no success unfortunately. 

Jax version: 0.11.2, CUDA13 on RTX5050

<img width="1914" height="960" alt="memory_usage" src="https://github.com/user-attachments/assets/099320a5-1113-4d6e-8fb3-a0c5edbb0f74" />

```
import jax
import matplotlib.pyplot as plt

DTYPE_INT     = jax.dtypes.canonicalize_dtype(int)

def single_point_lookup(
                input_data:     jax.Array,
                channel_offset: jax.Array,
                x:              jax.Array,
                y:              jax.Array,
                interp_type:    str) -> jax.Array:

    partial_tensor1     = jax.numpy.square(x + channel_offset)
    partial_tensor2     = jax.numpy.square(y + channel_offset)
    tot_indexing_tensor = jax.numpy.sqrt(partial_tensor1 + partial_tensor2)

    if interp_type == "nearest":
        out = input_data[jax.numpy.rint(tot_indexing_tensor).astype(DTYPE_INT)]

    elif interp_type == "linear":
        index_floor  = jax.numpy.trunc(tot_indexing_tensor).astype(DTYPE_INT)
        index_ceil   = index_floor + 1
        pixel_weight = tot_indexing_tensor - index_floor
        sample_indices = jax.numpy.stack((index_floor, index_ceil))
        samples = input_data[sample_indices]
        interpolation_weights = jax.numpy.stack((1 - pixel_weight, pixel_weight))
        out = jax.numpy.sum(samples * interpolation_weights)
    else:
        raise(ValueError("Wrong value for interp_type"))
    return out

@jax.jit(static_argnames=("interp_type")) #type: ignore [reportUnknownMemberType]
def data_grid_lookup(input_data: jax.Array,
                      channel_offset: jax.Array,
                      x: jax.Array,
                      y: jax.Array,
                      interp_type: str) -> jax.Array:

    #Grid x vmap
    multi_point_lookup_X = jax.vmap(single_point_lookup,(None, None, 0, None, None), 0)

    #Grid y vmap
    multi_point_lookup_XY = jax.vmap(multi_point_lookup_X, (None, None, None, 0, None), 0)

    #Channel batch which will be sum-reduced
    multi_point_lookup_XY_channel = jax.vmap( multi_point_lookup_XY, (0, 0,  None, None, None), 0)

    data_grid_out = jax.numpy.sum(multi_point_lookup_XY_channel( input_data, channel_offset, x, y, interp_type),axis=0)

    return data_grid_out

if __name__ == "__main__":
    Nchannel    = 500    #This axis will be reduced using sum
    Ndatapoints =  1800  #Axis to interpolate

    input_data     = jax.numpy.ones((Nchannel, Ndatapoints))
    channel_offset = jax.numpy.ones((Nchannel))

    memory_nearest_list = []
    memory_linear_list  = []

    dim_of_grid_list = range(20,180,10)

    for Npoints_per_grid_dim in dim_of_grid_list:
        x  = jax.numpy.ones((Npoints_per_grid_dim))
        y  = jax.numpy.ones((Npoints_per_grid_dim))

        input_tuple_recon_near = (input_data, channel_offset, x, y, "nearest" )
        input_tuple_recon_lin =  (input_data, channel_offset, x, y, "linear" )

        memory_nearest = data_grid_lookup.lower(*input_tuple_recon_near).compile().memory_analysis()

        total_memory_nearest = (
        memory_nearest.temp_size_in_bytes +
        memory_nearest.argument_size_in_bytes +
        memory_nearest.output_size_in_bytes -
        memory_nearest.alias_size_in_bytes
        )
        memory_nearest_list.append(total_memory_nearest)


        memory_linear  = data_grid_lookup.lower(*input_tuple_recon_lin).compile().memory_analysis()
        total_memory_linear = (
        memory_linear.temp_size_in_bytes +
        memory_linear.argument_size_in_bytes +
        memory_linear.output_size_in_bytes -
        memory_linear.alias_size_in_bytes
        )
        memory_linear_list.append(total_memory_linear)

    plt.figure
    plt.plot(dim_of_grid_list,jax.numpy.asarray(memory_nearest_list) * 1e-6)
    plt.plot(dim_of_grid_list,jax.numpy.asarray(memory_linear_list)  * 1e-6)
    plt.xlabel("Pixels per grid dimension in 2D")
    plt.ylabel("Total memory usage in MB")
    plt.legend(["Nearest Neighbour interp.", "Linear interp."])
    plt.show(block=True)
```
